### ISSUE TITLE
cluster setup: error logging podman into harbor

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonKUBE#1781]
I've been seeing cluster setup fail sometimes when it logs podman into harbor:

```
START-BUNDLE: ./script.sh
*** START TEXT FILE: script.sh
echo 'Vzfj0ZbevxVWlICRNbvV' | podman login registry.neon.local --username root --password-stdin
*** END TEXT FILE
*** START TEXT FILE: __run.sh
#!/bin/sh

chmod 700 "script.sh"
./script.sh
*** END TEXT FILE
*** Uploading: /home/sysadmin/.neon/exec/1dd780d7-e083-4337-9b92-d574f9c0a265/__bundle.zip
START: sudo bash -c 'if [ -f "/home/sysadmin/.neon/upload/__bundle.zip-30df4e23-d68a-45ae-bacc-0ace1a92b38e" ] ; then rm "/home/sysadmin/.neon/upload/__bundle.zip-30df4e23-d68a-45ae-bacc-0ace1a92b38e" ; fi'
END [OK]
STDERR
    Error: authenticating creds for "registry.neon.local": pinging container registry registry.neon.local: invalid status code from registry 503 (Service Unavailable)
END [ERROR=125]
END-BUNDLE
```

I saw this once or twice a week or so ago but am seeing it more often now, but not for every cluster setup.

I see that the code retries for up to 10 minutes in the setup step and then we retry the entire cluster setup operation for two more times, so we're giving harbor 30 minutes or more to stabilize, without success. 

I'm seeing a couple of distressed pods:

![Image](https://user-images.githubusercontent.com/4908431/237749284-7f1d69cc-47e4-4005-a3e0-0e69a5798f3b.png)

![Image](https://user-images.githubusercontent.com/4908431/237751013-00570c07-880f-4ff8-a2be-0c0824e859e3.png)

Here are the logs for the **neon-sso-oauth2proxy pod:***

```
λ kubectl logs -n neon-system neon-sso-oauth2-proxy-68cc4844cc-k5wwx
[2023/05/11 16:53:30] [main.go:75] WARNING: You are using alpha configuration. The structure in this configuration file may change without notice. You MUST remove conflicting options from your existing configuration.
[2023/05/11 16:53:30] [provider.go:55] Performing OIDC Discovery...
[2023/05/11 16:53:30] [main.go:60] ERROR: Failed to initialise OAuth2 Proxy: error intiailising provider: could not create provider data: error building OIDC ProviderVerifier: could not get verifier builder: error while discovery OIDC configuration: failed to discover OIDC configuration: error performing request: Get "https://neon-sso.e4cd-a747-528e-1b11.neoncluster.io/.well-known/openid-configuration": EOF
```

The **registry-harbor-harbor-core pod** never has the chance to start:

```
λ kubectl logs -n neon-system registry-harbor-harbor-core-748745798b-4s7l9
Error from server (BadRequest): container "core" in pod "registry-harbor-harbor-core-748745798b-4s7l9" is waiting to start: ContainerCreating
```

...and here is the description for the **registry-harbor-harbor-core pod:**

```
λ kubectl describe pod -n neon-system registry-harbor-harbor-core-748745798b-4s7l9                                                                                                                                                                                           
Name:                 registry-harbor-harbor-core-748745798b-4s7l9                                                                                                                                                                                                           
Namespace:            neon-system                                                                                                                                                                                                                                            
Priority:             900006000                                                                                                                                                                                                                                              
Priority Class Name:  neon-storage                                                                                                                                                                                                                                           
Service Account:      default                                                                                                                                                                                                                                                
Node:                 node/10.100.32.90                                                                                                                                                                                                                                      
Start Time:           Thu, 11 May 2023 08:38:15 -0700                                                                                                                                                                                                                        
Labels:               core.goharbor.io/name=registry-harbor-harbor-core                                                                                                                                                                                                      
                      core.goharbor.io/namespace=neon-system                                                                                                                                                                                                                 
                      goharbor.io/operator-controller=core                                                                                                                                                                                                                   
                      goharbor.io/operator-version=v1.3.0                                                                                                                                                                                                                    
                      pod-template-hash=748745798b                                                                                                                                                                                                                           
Annotations:          goharbor.io/warning: ⚠️ This Resource is managed by *core* ⚠️                                                                                                                                                                                          
                      registry-harbor-harbor-core.neon-system.configmap.core.goharbor.io/uuid: 6457da05-9e92-4f73-b796-803dfa20e96f                                                                                                                                          
                      registry-harbor-harbor-core.neon-system.configmap.core.goharbor.io/version: 5000                                                                                                                                                                       
                      registry-harbor-harbor-core.neon-system.secret.core.goharbor.io/uuid: 11e5eaa4-5c8a-4215-bac7-e339f0f8d995                                                                                                                                             
                      registry-harbor-harbor-core.neon-system.secret.core.goharbor.io/version: 4999                                                                                                                                                                          
                      sidecar.istio.io/inject: false                                                                                                                                                                                                                         
Status:               Pending                                                                                                                                                                                                                                                
IP:                                                                                                                                                                                                                                                                          
IPs:                  <none>                                                                                                                                                                                                                                                 
Controlled By:        ReplicaSet/registry-harbor-harbor-core-748745798b                                                                                                                                                                                                      
Containers:                                                                                                                                                                                                                                                                  
  core:                                                                                                                                                                                                                                                                      
    Container ID:                                                                                                                                                                                                                                                            
    Image:          registry.neon.local/neonkube/harbor-core:v2.5.2                                                                                                                                                                                                          
    Image ID:                                                                                                                                                                                                                                                                
    Port:           8080/TCP                                                                                                                                                                                                                                                 
    Host Port:      0/TCP                                                                                                                                                                                                                                                    
    State:          Waiting                                                                                                                                                                                                                                                  
      Reason:       ContainerCreating                                                                                                                                                                                                                                        
    Ready:          False                                                                                                                                                                                                                                                    
    Restart Count:  0                                                                                                                                                                                                                                                        
    Liveness:       http-get http://:http/api/v2.0/ping delay=0s timeout=1s period=90s #success=1 #failure=3                                                                                                                                                                 
    Readiness:      http-get http://:http/api/v2.0/ping delay=0s timeout=1s period=10s #success=1 #failure=3                                                                                                                                                                 
    Environment:                                                                                                                                                                                                                                                             
      WITH_CHARTMUSEUM:                          false                                                                                                                                                                                                                       
      AUTH_MODE:                                 db_auth                                                                                                                                                                                                                     
      DATABASE_TYPE:                             postgresql                                                                                                                                                                                                                  
      CORE_URL:                                  http://registry-harbor-harbor-core:80                                                                                                                                                                                       
      JOBSERVICE_URL:                            http://registry-harbor-harbor-jobservice                                                                                                                                                                                    
      WITH_TRIVY:                                false                                                                                                                                                                                                                       
      EXT_ENDPOINT:                              https://neon-registry.e4cd-a747-528e-1b11.neoncluster.io                                                                                                                                                                    
      POSTGRESQL_PASSWORD:                       <set to the key 'postgresql-password' in secret 'registry'>  Optional: false                                                                                                                                                
      POSTGRESQL_USERNAME:                       neon_service                                                                                                                                                                                                                
      POSTGRESQL_DATABASE:                       harbor_core                                                                                                                                                                                                                 
      CORE_LOCAL_URL:                            http://registry-harbor-harbor-core:80                                                                                                                                                                                       
      REGISTRY_CONTROLLER_URL:                   http://registry-harbor-harbor-registryctl                                                                                                                                                                                   
      POSTGRESQL_HOST:                           neon-system-db.neon-system                                                                                                                                                                                                  
      POSTGRESQL_PORT:                           5432                                                                                                                                                                                                                        
      HARBOR_ADMIN_PASSWORD:                     <set to the key 'secret' in secret 'registry'>  Optional: false                                                                                                                                                             
      WITH_NOTARY:                               false                                                                                                                                                                                                                       
      REGISTRY_URL:                              http://registry-harbor-harbor-registry                                                                                                                                                                                      
      TOKEN_SERVICE_URL:                         http://registry-harbor-harbor-core/service/token                                                                                                                                                                            
      METRIC_ENABLE:                             false                                                                                                                                                                                                                       
      TRACE_ENABLED:                             true                                                                                                                                                                                                                        
      TRACE_SERVICE_NAME:                        harbor-core                                                                                                                                                                                                                 
      TRACE_OTEL_TIMEOUT:                        10                                                                                                                                                                                                                          
      TRACE_OTEL_COMPRESSION:                    false                                                                                                                                                                                                                       
      TRACE_OTEL_INSECURE:                       true                                                                                                                                                                                                                        
      TRACE_SAMPLE_RATE:                         1                                                                                                                                                                                                                           
      TRACE_NAMESPACE:                                                                                                                                                                                                                                                       
      TRACE_OTEL_ENDPOINT:                       grafana-agent-node.neon-monitor.svc.cluster.local:4318                                                                                                                                                                      
      TRACE_OTEL_URL_PATH:                       /v1/traces                                                                                                                                                                                                                  
      LOG_LEVEL:                                 info                                                                                                                                                                                                                        
      CORE_SECRET:                               <set to the key 'secret' in secret 'registry-harbor-harbor-core-secret'>    Optional: false                                                                                                                                 
      _REDIS_URL_CORE:                           <set to the key '_REDIS_URL_CORE' in secret 'registry-harbor-harbor-core'>  Optional: false                                                                                                                                 
      PORTAL_URL:                                http://registry-harbor-harbor-portal                                                                                                                                                                                        
      REGISTRY_CREDENTIAL_USERNAME:              harbor_registry_user                                                                                                                                                                                                        
      REGISTRY_CREDENTIAL_PASSWORD:              <set to the key 'secret' in secret 'registry-harbor-harbor-registry-basicauth'>  Optional: false                                                                                                                            
      JOBSERVICE_SECRET:                         <set to the key 'secret' in secret 'registry-harbor-harbor-jobservice-secret'>   Optional: false                                                                                                                            
      CSRF_KEY:                                  <set to the key 'key' in secret 'registry-harbor-harbor-core-csrf'>              Optional: false                                                                                                                            
      CONFIG_PATH:                               /etc/core/app.conf                                                                                                                                                                                                          
      CFG_EXPIRATION:                            5                                                                                                                                                                                                                           
      RELOAD_KEY:                                true                                                                                                                                                                                                                        
      SYNC_QUOTA:                                true                                                                                                                                                                                                                        
      SYNC_REGISTRY:                             false                                                                                                                                                                                                                       
      INTERNAL_TLS_ENABLED:                      false                                                                                                                                                                                                                       
      PERMITTED_REGISTRY_TYPES_FOR_PROXY_CACHE:  docker-hub,harbor,azure-acr,aws-ecr,google-gcr,quay,docker-registry                                                                                                                                                         
      POSTGRESQL_MAX_IDLE_CONNS:                 50                                                                                                                                                                                                                          
      POSTGRESQL_MAX_OPEN_CONNS:                 1000                                                                                                                                                                                                                        
      POSTGRESQL_SSLMODE:                        disable                                                                                                                                                                                                                     
      _REDIS_URL_REG:                            <set to the key '_REDIS_URL_REG' in secret 'registry-harbor-harbor-core'>  Optional: false                                                                                                                                  
    Mounts:                                                                                                                                                                                                                                                                  
      /etc/core/app.conf from configuration (ro,path="app.conf")                                                                                                                                                                                                             
      /etc/core/ca from ca-download (rw)                                                                                                                                                                                                                                     
      /etc/core/key from encryption (ro,path="key")                                                                                                                                                                                                                          
      /etc/core/private_key.pem from token-service-private-key (ro,path="tls.key")                                                                                                                                                                                           
      /etc/core/token from psc (rw)                                                                                                                                                                                                                                          
      /harbor_cust_cert from internal-certificates (rw)                                                                                                                                                                                                                      
Conditions:                                                                                                                                                                                                                                                                  
  Type              Status                                                                                                                                                                                                                                                   
  Initialized       True                                                                                                                                                                                                                                                     
  Ready             False                                                                                                                                                                                                                                                    
  ContainersReady   False                                                                                                                                                                                                                                                    
  PodScheduled      True                                                                                                                                                                                                                                                     
Volumes:                                                                                                                                                                                                                                                                     
  configuration:                                                                                                                                                                                                                                                             
    Type:      ConfigMap (a volume populated by a ConfigMap)                                                                                                                                                                                                                 
    Name:      registry-harbor-harbor-core                                                                                                                                                                                                                                   
    Optional:  false                                                                                                                                                                                                                                                         
  encryption:                                                                                                                                                                                                                                                                
    Type:        Secret (a volume populated by a Secret)                                                                                                                                                                                                                     
    SecretName:  registry-harbor-harbor-core-encryptionkey                                                                                                                                                                                                                   
    Optional:    false                                                                                                                                                                                                                                                       
  psc:                                                                                                                                                                                                                                                                       
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)                                                                                                                                                                                                
    Medium:                                                                                                                                                                                                                                                                  
    SizeLimit:  <unset>                                                                                                                                                                                                                                                      
  token-service-private-key:                                                                                                                                                                                                                                                 
    Type:        Secret (a volume populated by a Secret)                                                                                                                                                                                                                     
    SecretName:  registry-harbor-harbor-core-tokencert                                                                                                                                                                                                                       
    Optional:    false                                                                                                                                                                                                                                                       
  ca-download:                                                                                                                                                                                                                                                               
    Type:        Secret (a volume populated by a Secret)                                                                                                                                                                                                                     
    SecretName:  neon-cluster-certificate                                                                                                                                                                                                                                    
    Optional:    false                                                                                                                                                                                                                                                       
  internal-certificates:                                                                                                                                                                                                                                                     
    Type:        EmptyDir (a temporary directory that shares a pod's lifetime)                                                                                                                                                                                               
    Medium:                                                                                                                                                                                                                                                                  
    SizeLimit:   <unset>                                                                                                                                                                                                                                                     
QoS Class:       BestEffort                                                                                                                                                                                                                                                  
Node-Selectors:  neonkube.io/neon-system.registry=true                                                                                                                                                                                                                       
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 30s                                                                                                                                                                                                    
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 30s                                                                                                                                                                                                  
Events:                                                                                                                                                                                                                                                                      
  Type     Reason       Age                   From     Message                                                                                                                                                                                                               
  ----     ------       ----                  ----     -------                                                                                                                                                                                                               
  Warning  FailedMount  40m (x4 over 79m)     kubelet  Unable to attach or mount volumes: unmounted volumes=[ca-download], unattached volumes=[psc token-service-private-key ca-download internal-certificates configuration encryption]: timed out waiting for the condition
  Warning  FailedMount  10m (x7 over 60m)     kubelet  Unable to attach or mount volumes: unmounted volumes=[ca-download], unattached volumes=[token-service-private-key ca-download internal-certificates configuration encryption psc]: timed out waiting for the condition
  Warning  FailedMount  5m42s (x45 over 81m)  kubelet  MountVolume.SetUp failed for volume "ca-download" : secret "neon-cluster-certificate" not found                                                                                                                       
```

**Comment by jefflill** [comment]
@marcusbooyah: I think you'll need to have a look at this.

**Comment by jefflill** [comment]
**never mind:** I F5'd to complete the unit test and forgot that the test deletes the cluster when it's done, so this cluster isn't present any more.

I'm using the build machine right now to rebuild the node and desktop images and then I'll try recreating the problem.

**Comment by jefflill** [comment]
I wonder if the underlying secret volume mount issue is happening because the secrets are being created in parallel with starting the Harbor deployments.

Could it be possible that **sometimes the secrets don't exist when pods are started** resulting in the **secret mount failures** (which apparently, the pod/deployment can't recover from)?

**Comment by jefflill** [comment]
I think this may be due to a change I made to  *KubernetesRetryHandler** a few weeks ago where I dramatically reduced the retry timeout.  I've  since changed that to be more configurable and restored the default time out to be much longer that what I changed it to but not as long as it was original (60min if I remember correctly).

I'm going to do a few hundred cluster deployments to torture test this.
